### PR TITLE
SF-1271: Include sessionID as a part of all requests

### PR DESCRIPTION
**Pull request by JoeyDeol**

### CODE REVIEW

File: packages/@storefront/flux-capacitor/test/unit/core/reducers/session.ts [hunk 1]
```diff
@@ -118,4 +118,17 @@ suite('session', ({ expect }) => {
...
   });
+
+  describe('setSessionId', () => {
+    it('should update state with sessionId', () => {
+      const sessionId = 'foo';
+      const newState = {
+        ...state,
+        sessionId
+      };
+      const reducer = session(state, <any>{ type: Actions.SET_SESSION_ID, payload: sessionId });
```
Comment (jrmykolyn) [inline packages/@storefront/flux-capacitor/test/unit/core/reducers/session.ts]: Newline between 'arrange' and 'assert' stages.

File: packages/@storefront/flux-capacitor/src/core/requests/utils.ts [hunk 2]
```diff
@@ -182,6 +182,11 @@ namespace RequestHelpers {
 
   export const chain = <T>(...fns: Array<(...obj: any[]) => T>): T =>
     fns.reduce((final, fn) => fn(final) || final, <T>{});
+
+  export const attachSessionId = <T>(state: Store.State): ((r: T) => T&{ sessionId: Store.SessionId }) => {
+    const sessionId = state.session.sessionId;
```
Comment (jrmykolyn) [inline packages/@storefront/flux-capacitor/src/core/requests/utils.ts]: Unless you're going to set a fallback, just inline this.

ie.

```
...{ sessionId: state.session.sessionId }...
```

File: packages/@storefront/flux-capacitor/src/core/requests/utils.ts [hunk 3]
```diff
@@ -182,6 +182,11 @@ namespace RequestHelpers {
 
   export const chain = <T>(...fns: Array<(...obj: any[]) => T>): T =>
     fns.reduce((final, fn) => fn(final) || final, <T>{});
+
+  export const attachSessionId = <T>(state: Store.State): ((r: T) => T&{ sessionId: Store.SessionId }) => {
```
Comment (jrmykolyn) [inline packages/@storefront/flux-capacitor/src/core/requests/utils.ts]: re: `T&{ ...`, do we normally include a space between the `&` operator and operands? If so, let's do that here.

File: packages/@storefront/flux-capacitor/src/core/actions/creators.ts [hunk 4]
```diff
@@ -1159,6 +1159,15 @@ namespace ActionCreators {
     return createAction(Actions.UPDATE_LOCATION, location);
   }
 
+  /**
+   * Set the sessionId in the store to the given id.
+   * @param sessionID- The sessionId to set.
```
Comment (jrmykolyn) [inline packages/@storefront/flux-capacitor/src/core/actions/creators.ts]: Misc.
- Param name should be `sessionId` (not `sessionID`).
- Space between `sessionId` and trailing `-`.

File: packages/@storefront/flux-capacitor/CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,13 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased] [minor]
+### Changed
+- SF-1271: The `composeRequest` function now uses the `attachSessionId` function to add sessionIds to every StoreFront request.
```
Comment (jrmykolyn) [inline packages/@storefront/flux-capacitor/CHANGELOG.md]: Include trailing `()` to signify that the snippet is a function. Drop 'function' word.

File: packages/@storefront/core/test/unit/services/tracker.ts [hunk 6]
```diff
@@ -59,11 +59,23 @@ suite('Tracker Service', ({ expect, spy, stub, itShouldExtendBaseService }) => {
   });
 
   describe('init()', () => {
+    const sessionId = 'foo';
+    const SET_SESSION_ID = 'SET_SESSION_ID';
+
+    beforeEach(() => {
+      app.flux = <any> {
+        on,
+        store: { dispatch: spy() },
```
Comment (jrmykolyn) [inline packages/@storefront/core/test/unit/services/tracker.ts]: Capture the dispatch-specific `spy` in a variable (ie. `dispatch`); make assertions on `dispatch` directly.

File: packages/@storefront/core/test/unit/services/tracker.ts [hunk 7]
```diff
@@ -72,13 +84,24 @@ suite('Tracker Service', ({ expect, spy, stub, itShouldExtendBaseService }) => {
...
     });
+
+    it('should dispatch a setSessionId action', () => {
+      const disableWarnings = spy();
+      const getSessionId = spy();
+      service.client = <any>{ disableWarnings, getSessionId, autoSetVisitor: () => null };
+
+      service.init();
+
+      expect(app.flux.store.dispatch).to.be.calledWith(SET_SESSION_ID);
```
Comment (jrmykolyn) [inline packages/@storefront/core/test/unit/services/tracker.ts]: Update assertion to reference `dispatch` directly (see note above).

File: packages/@storefront/flux-capacitor/test/unit/core/requests/utils.ts [hunk 8]
```diff
@@ -572,12 +572,24 @@ suite('requests helpers', ({ expect, stub, spy }) => {
...
       expect(RequestHelpers.chain(
         utils.normalizeToFunction(<any>{ a: 'b' }),
         (x) => null,
         utils.normalizeToFunction({ e: 'f' })
       )).to.eql({ a: 'b', e: 'f' });
     });
   });
+
+  describe('attachSessionId()', () => {
+    it('should return a function that is called with the given request and add sessionId to the request', () => {
```
Comment (jrmykolyn) [inline packages/@storefront/flux-capacitor/test/unit/core/requests/utils.ts]: Simplify to something like: "it should return a function which applies the session ID to the request".

File: packages/@storefront/flux-capacitor/test/unit/core/requests/utils.ts [hunk 9]
```diff
@@ -572,12 +572,24 @@ suite('requests helpers', ({ expect, stub, spy }) => {
...
+      const req = { a: 'b' };
+      const sessionId = 'foo';
+      const state: any = { c: 'd', session: { sessionId } };
+
+      const requestWithSessionId = RequestHelpers.attachSessionId(state)(req);
```
Comment (jrmykolyn) [inline packages/@storefront/flux-capacitor/test/unit/core/requests/utils.ts]: Consider splitting this test into multiple assertions. Something that would:

- Invoke `attachSessionId`, capture the result, and assert that it's a function.
- Invoke the resulting function with the base request, capture the result, and assert that it contains the additional data.

File: packages/@storefront/flux-capacitor/src/core/requests/utils.ts (hunk 3 above)
Comment (JoeyDeol) [reply packages/@storefront/flux-capacitor/src/core/requests/utils.ts]: Will do.

File: packages/@storefront/flux-capacitor/src/core/actions/creators.ts (hunk 4 above)
Comment (JoeyDeol) [reply packages/@storefront/flux-capacitor/src/core/actions/creators.ts]: Done.

File: packages/@storefront/flux-capacitor/CHANGELOG.md (hunk 5 above)
Comment (JoeyDeol) [reply packages/@storefront/flux-capacitor/CHANGELOG.md]: Will change.

File: packages/@storefront/flux-capacitor/src/core/requests/utils.ts (hunk 2 above)
Comment (JoeyDeol) [reply packages/@storefront/flux-capacitor/src/core/requests/utils.ts]: Done.

File: packages/@storefront/core/test/unit/services/tracker.ts (hunk 6 above)
Comment (JoeyDeol) [reply packages/@storefront/core/test/unit/services/tracker.ts]: Done.

File: packages/@storefront/core/test/unit/services/tracker.ts (hunk 7 above)
Comment (JoeyDeol) [reply packages/@storefront/core/test/unit/services/tracker.ts]: Done.

File: packages/@storefront/flux-capacitor/test/unit/core/requests/utils.ts (hunk 9 above)
Comment (JoeyDeol) [reply packages/@storefront/flux-capacitor/test/unit/core/requests/utils.ts]: Done.

**Pull request merged**